src/controllers/networks/networks.ts

```diff
@@ -0,0 +1,278 @@
+  #networks: { [key: NetworkId]: Network } = {}
+
+  statuses: Statuses<keyof typeof STATUS_WRAPPED_METHODS> = STATUS_WRAPPED_METHODS
+
+  networkToAddOrUpdate: {
+    chainId: Network['chainId']
+    rpcUrl: string
+    info?: NetworkInfoLoading<NetworkInfo>
+  } | null = null
+
+  #onRemoveNetwork: (id: NetworkId) => void
+
+  #onAddOrUpdateNetwork: (network: Network) => void
+
+  // Holds the initial load promise, so that one can wait until it completes
+  initialLoadPromise: Promise<void>
+
+  constructor(
+    storage: Storage,
+    onAddOrUpdateNetwork: (network: Network) => void,
+    onRemoveNetwork: (id: NetworkId) => void
+  ) {
+    super()
+    this.#storage = storage
+    this.#onAddOrUpdateNetwork = onAddOrUpdateNetwork
+    this.#onRemoveNetwork = onRemoveNetwork
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.initialLoadPromise = this.#load()
+  }
+
+  get isInitialized(): boolean {
+    return !!Object.keys(this.#networks).length
+  }
+
+  get networks(): Network[] {
+    if (!this.#networks) return predefinedNetworks
+
+    const uniqueNetworksByChainId = Object.values(this.#networks)
+      .sort((a, b) => +b.predefined - +a.predefined) // first predefined
+      .filter((item, index, self) => self.findIndex((i) => i.chainId === item.chainId) === index) // unique by chainId (predefined with priority)
+    return uniqueNetworksByChainId.map((network) => {
+      // eslint-disable-next-line no-param-reassign
+      network.features = getFeaturesByNetworkProperties({
+        isSAEnabled: network.isSAEnabled,
+        isOptimistic: network.isOptimistic ?? false,
+        rpcNoStateOverride: network.rpcNoStateOverride,
+        erc4337: network.erc4337,
+        areContractsDeployed: network.areContractsDeployed,
+        feeOptions: network.feeOptions,
+        hasDebugTraceCall: network.hasDebugTraceCall,
+        platformId: network.platformId,
+        nativeAssetId: network.nativeAssetId,
+        flagged: network.flagged ?? false,
+        chainId: network.chainId,
+        hasSingleton: network.hasSingleton
+      })
+      return network
+    })
+  }
+
+  async #load() {
+    const storedNetworkPreferences: { [key: NetworkId]: Partial<Network> } | undefined =
+      await this.#storage.get('networkPreferences', undefined)
+    let storedNetworks: { [key: NetworkId]: Network }
+    storedNetworks = await this.#storage.get('networks', undefined)
+    if (!storedNetworks && storedNetworkPreferences) {
+      storedNetworks = await migrateNetworkPreferencesToNetworks(storedNetworkPreferences)
+      await this.#storage.set('networks', storedNetworks)
+      await this.#storage.remove('networkPreferences')
+    }
+    if (!storedNetworks) {
+      storedNetworks = predefinedNetworks.reduce((acc, n) => {
+        acc[n.id] = n
+        return acc
+      }, {} as { [key: NetworkId]: Network })
+      await this.#storage.set('networks', storedNetworks)
+    }
+    this.#networks = storedNetworks
```

Comment:
If we add or remove a predefined network, we should also update the networks stored in the storage. We should probably consider using deep equal, since if we add, remove or update attr to an already existing network, this change won't go in as well if we only search for diff (in terms of new ones).